src/ipc/ipc-common/IPCProxy.h

```diff
@@ -61,45 +61,48 @@ class IPCProxy : public WrapperType, public IPCProxyNewBase
     {
         QObject::connect(ipc(), &IPCProxyBinderBase::complete, this, [this] () {
                 m_localProviderBinder.init();
-                if (m_localProviderBinder.provider() == nullptr) {
-                    auto serviceName = ipc()->serviceName();
-                    auto objPath = ipc()->objectPath();
-                    for (auto& proxy : m_ipcProxies) {
-                        auto proxyAdapterIPCBinder = proxy.ipcBinder;
-                        if (proxyAdapterIPCBinder != nullptr) {
-                            proxyAdapterIPCBinder->setObjectPath(objPath);
-                            if(!serviceName.isEmpty()) {
-                                proxyAdapterIPCBinder->setServiceName(serviceName);
-                            }
-                            proxyAdapterIPCBinder->connectToServer();
-                            QObject::connect(proxyAdapterIPCBinder, &IPCProxyBinderBase::serviceAvailableChanged, this, &IPCProxy::refreshProvider);
-                        }
+                for (auto& proxy : m_ipcProxies) {
+                    auto ipcBinder = proxy.ipcBinder;
+                    if (ipcBinder != nullptr) {
+                        QObject::connect(ipcBinder, &IPCProxyBinderBase::serviceAvailableChanged, this, &IPCProxy::refreshProvider);
                     }
                 }
                 this->refreshProvider();
-            });
+        });
+    }
 
+    void enableBinder(IPCProxyBinderBase * ipcBinder)
+    {
+        ipcBinder->setObjectPath(ipc()->objectPath());
+        const auto &serviceName = ipc()->serviceName();
+        if (!serviceName.isEmpty()) {
+            ipcBinder->setServiceName(serviceName);
+        }
+        ipcBinder->connectToServer();
     }
 
     void refreshProvider() override
     {
         Q_ASSERT(m_defaultProvider);
         InterfaceType *provider = m_defaultProvider;
+        bool enableFollowingProxies = true;
         if (m_localProviderBinder.provider() != nullptr) {
             provider = m_localProviderBinder.provider();
-            for (auto& proxy : m_ipcProxies) {
-                auto ipcBinder = proxy.ipcBinder;
-                if (ipcBinder != nullptr) {
-                    ipcBinder->setEnabled(false);
-                }
-            }
-        } else {
-            for (auto& proxy : m_ipcProxies) {
-                auto ipcBinder = proxy.ipcBinder;
-                if ((ipcBinder != nullptr) && ipcBinder->isServiceAvailable()) {
-                    provider = proxy.proxy;
-                    ipcBinder->setEnabled(true);
-                    break;
+            enableFollowingProxies = false;
+        }
+
+        for (auto& proxy : m_ipcProxies) {
+            auto ipcBinder = proxy.ipcBinder;
+            if (ipcBinder != nullptr) {
+                ipcBinder->setEnabled(enableFollowingProxies);
+
+                if (enableFollowingProxies) {
```

Comment:
It won't call setObjectPath on every ipcBinder

Reply:
you are right, and I seen nothing wrong here.